### ISSUE TITLE
[BUG]: There seems to be a memory leak

### ISSUE THREAD

**Issue by mvdwetering** [plugwise/plugwise_usb-beta#45]
### Describe the bug.

I have been running into issues with HA memory usage keep growing until it eventually hangs HA. I finally found that when stopping this integration that the memory increase stops. So it seems it is related to this integration.

As an example see the picture below. The blue line in HA memory % usage. I stopped this integration on 16th of July around 13:00 after which it stabilizes. Before that you can see a few spikes where memory usage starts rising. I just restart HA to avoid it causing problems so that is why the spikes go down.
![image](https://github.com/plugwise/plugwise_usb-beta/assets/732514/2bba38c2-2595-4301-abef-86cbfe77a5c3)

### What version of Plugwise Beta are you using?

v0.40.1

### What version of Home Assistant Core are you using?

2023.7.3

### What type of installation are you running?

Home Assistant OS

### How did you install plugwise-beta?

HACS

### What kind of Plugwise device is having issues?

USB: Stick

### What firmware version is your Plugwise product at?

Circle+ 2011-06-27 10:47:37, Circle 2011-06-27 10:52:18, 8x Circle  2009-09-08 15:50:31

### Logging

I enabled logging with the "Enable debug logging" option in the integration.
Will add it after a few hours of grabbing logging.

I don't really want to lose the info I typed in the bug report until now.

### Diagnostics information

I don't see the diagnostics download option for this integration. Do see it for others.

### Additional information

Might have started after I added 8 more plugs to the network a while back. Don't think I noticed issues before.

**Comment by mvdwetering** [comment]
I have run the log_objects service from the Home Assistant profiler integration for a while. I attached the logs. The Memory Growth indicates the `(object, count, increase since last time)`
[profiler.log](https://github.com/plugwise/plugwise_usb-beta/files/12329577/profiler.log)

It seems that for some reason at some point the integration starts creating more and more `LogAddr` and `CircleEnergyCountersRequest` objects. At the end of the log it is about 1900 new ones per objecttype per 30 seconds, which adds up. At the end of my log it had over 2 million of those objects.

It is unclear to me why the integration seems to work for a while and then it starts piling up.


Here is also part of the home-assistant.log from where I started the profiler. Memory was already increasing at a steady pace at that point,. Only posting part since the whole log is about 1.7GB.
[partial.log](https://github.com/plugwise/plugwise_usb-beta/files/12329605/partial.log)

The things that stand out most are:
* A lot of `_collect_energy_pulses for 000D6F000072A230 at 2023-08-12 07:00:00 not found, request counter from memory` entries
* A lot of `Drop unexpected success <command> using seq_id b'<someid>`

For the last one I get the impression that sequenceids are confused/messed up. That is assuming they are supposed to be sequential and unique per message.

From the short log fragement below it looks like it got 2 responses for different commands with sequenceid `b'939F'` which should not happen if the sequenceid is supposed to be unique. Next to that it looks like it is sending messages with sequence `b'FFF3'` which is a huge difference from 939F. 

Not sure what to think of this, but maybe it rings a bell with someone more knowledgeable about the implementation.

```
2023-08-12 14:51:34.975 INFO (send_messages_thread) [plugwise_usb.controller] Resend CircleEnergyCountersRequest for 000D6F000059B01F, retry 1 of 3
2023-08-12 14:51:34.975 INFO (send_messages_thread) [plugwise_usb.controller] Resend CircleEnergyCountersRequest to b'000D6F000059B01F' using seq_id b'FFF3', retry 1
2023-08-12 14:51:35.003 INFO (serial_reader_thread) [plugwise_usb.controller] Drop unexpected success NodeAckSmallResponse using seq_id b'939F'
2023-08-12 14:51:35.063 INFO (serial_reader_thread) [plugwise_usb.controller] Received CircleEnergyCountersResponse from 000D6F000059B01F with sequence id b'939F'
2023-08-12 14:51:35.063 INFO (serial_reader_thread) [plugwise_usb.nodes.circle] _response_energy_counters for 000D6F000059B01F | collection not running, len=80, timestamp:2023-08-11 17:00:00=2023-08-12 12:00:00
2023-08-12 14:51:35.063 INFO (serial_reader_thread) [plugwise_usb.nodes.circle] _collect_energy_pulses for 000D6F000059B01F at 2023-08-12 12:00:00 not found, request counter from memory 1678 (from mem=1678, slot=0, timestamp=2023-08-11 17:00:00)
2023-08-12 14:51:35.063 INFO (serial_reader_thread) [plugwise_usb.nodes.circle] _update_energy_previous_hour for 000D6F000059B01F | counter = None, update= None, timestamp 2023-08-12 12:00:00
2023-08-12 14:51:35.063 INFO (serial_reader_thread) [plugwise_usb.nodes.circle] _collect_energy_pulses for 000D6F000059B01F at 2023-08-11 23:00:00 not found, request counter from memory 1675 (from mem=1678, slot=0, timestamp=2023-08-11 17:00:00)
<more collect energy lines>
2023-08-12 14:51:35.064 INFO (serial_reader_thread) [plugwise_usb.nodes.circle] _update_energy_today_hourly for 000D6F000059B01F | counter = None, update= None, range 2023-08-11 23:00:00 to 2023-08-12 12:00:00
<more collect energy lines>
2023-08-12 14:51:35.065 INFO (serial_reader_thread) [plugwise_usb.nodes.circle] _update_energy_today_now for 000D6F000059B01F | skip update, hour: False=True=False, history: False=False=True, day: False=False=True
2023-08-12 14:51:35.065 INFO (serial_reader_thread) [plugwise_usb.controller] Drop unexpected  CircleEnergyCountersResponse using seq_id b'939F'
2023-08-12 14:51:35.503 CRITICAL (SyncWorker_6) [homeassistant.components.profiler] Memory Growth: [('list', 2263849, 494), ('LogAddr', 845556, 197), ('CircleEnergyCountersRequest', 845554, 197), ('CirclePowerUsageRequest', 111502, 20), ('NodePingRequest', 139140, 20), ('ReadOnlyDict', 13858, 9), ('Event', 1578, 4), ('State', 2857, 3), ('MemoryBIO', 12, 2), ('_SSLSocket', 13, 1), ('_SSLProtocolTransport', 5, 1), ('SSLProtocol', 5, 1), ('SSLObject', 6, 1), ('HttpResponseParser', 6, 1), ('StreamReader', 7, 1)]
2023-08-12 14:51:36.205 INFO (send_messages_thread) [plugwise_usb.controller] Resend CircleEnergyCountersRequest for 000D6F000059B01F, retry 2 of 3
2023-08-12 14:51:36.205 INFO (send_messages_thread) [plugwise_usb.controller] Resend NodePingRequest to b'000D6F0000728DFD' using seq_id b'FFF3', retry 1
```


If you need more specific logging I can generate that.

**Comment by CoMPaTech** [comment]
That looks like a good catch of what might be going (wr)on(g), I'll try to look at the code this week if I can spot something depicting that spiral path. Thanks for the extended info!

**Comment by CoMPaTech** [comment]
Most probably coming from the module, can't find the second instance (yet) but the first logging instance comes from https://github.com/plugwise/python-plugwise-usb/blob/e419428383aa54a2f22ced29d67ead924a6905de/plugwise_usb/nodes/circle.py#L610

**Comment by Herbus67** [comment]
![Scherm­afbeelding 2023-09-26 om 08 57 57](https://github.com/plugwise/plugwise_usb-beta/assets/133803463/1f137b9d-9c46-49c6-a3a4-3bae8254a547)
I've the same issue. Turned off the integration to find out if this was causing the memory usage increase. When turned off it was stable for weeks, after turning on it's increasing rapidly.

**Comment by dirixmjm** [comment]
Most likely this is fixed as well in 0.31.1b0.
See https://github.com/plugwise/plugwise_usb-beta/discussions/110

**Comment by bouwew** [comment]
@mvdwetering @Herbus67 please test v0.40.4 and let us know your result.

**Comment by mvdwetering** [comment]
Thanks, I installed it just now. Will need to run for some days until I can see if it helped.

**Comment by dirixmjm** [comment]
@mvdwetering I have found another underlying issue so probably you will still have some issues with the last update. I hope to be able to release another beta version soon. If you are willing to help with testing please join the conversation at https://github.com/plugwise/plugwise_usb-beta/discussions/110

**Comment by mvdwetering** [comment]
Yeah, I am subscribed to that one, but there is a lot happening and a bit hard to follow :)
I will try to setup the integration in my dev environment so I can try the changes in the library as my main machine is running HA OS where that is complicated. If I can get that working I will report in the other topic.

**Comment by dirixmjm** [comment]
Ok don't worry, there are currently three testers (including myself). On my testing environment it is stable for the the last hours, if that continues we can make another update to the component.

**Comment by dirixmjm** [comment]
@mvdwetering would you be up for doing some beta-testing?

There is a new release 0.31.2b3 of the python-plugwise-usb backend.
To do the testing make sure you have upgraded plugwise-usb-beta to 40.4 (inside HACS). Then inside the HA configuration folder you edit the file custom_components/plugwise_usb/manifest.json and set the parameter:
  "requirements": ["plugwise-usb==0.31.2b2"],

**Comment by mvdwetering** [comment]
I was already running 40.4, but noticed that memory started climbing again.

Just updated the manifest and restarted HA. Will keep en eye on it.

**Comment by dirixmjm** [comment]
@mvdwetering alpha version 40.5a1 has been released, can you start testing?

**Comment by mvdwetering** [comment]
I just updated.

Unfortunately did not get results with the previous version as I had some other issues that required several restarts of HA so it never ran for a long enough time.

**Comment by dirixmjm** [comment]
Hi @mvdwetering version v0.40.5 was just released. If you could test with this version that would be appreciated.

**Comment by mvdwetering** [comment]
I just updated.

The previous version seemed better, but hard to say as I had to restart HA  afew times so no very long uptime.

**Comment by dirixmjm** [comment]
In what terms better?

**Comment by mvdwetering** [comment]
Previously there was an exponential memory increase after some days and claiming all memory. See 2 blue spikes in image and first post. Note that I have notifications on high-ish memory usage so I can check the memory increase and restart HA, so that is why the peaks don't seem that high, but leaving it running for a few hours more would make it consume all memory.

![image](https://github.com/plugwise/plugwise_usb-beta/assets/732514/44d0daa9-e73e-4cc5-b7c4-52180ca7cbed)

It seems that something is still claiming memory, but it seems at a slower/steadier pace, which is why I called it better. It could even be another integration, but have to wait until it starts rising again so I can disable integrations to see when it stops.

I will replug the stick and restart HA to have a clean start now.

**Comment by mvdwetering** [comment]
My system has been running for almost 2 weeks now with v0.40.5 without having the exponential memory increases anymore. These increases would start after a few days of running, so 2 weeks seems long enough.

See lightblue/highest line which shows HA core memory usage in the chart below. The system was restarted at the drop between 22 and 24 and memory usage has been mostly stable, except for some weird step on 26th, but I don't know yet which integration caused that.

![image](https://github.com/plugwise/plugwise_usb-beta/assets/732514/ccc2b79b-17ed-4fe8-8926-af326b126c60)

So the issue seems to be fixed.
